src/app/icd/ICDManager.cpp

```diff
@@ -79,12 +83,65 @@ bool ICDManager::SupportsCheckInProtocol()
     bool success        = false;
     uint32_t featureMap = 0;
     // Can't use attribute accessors/Attributes::FeatureMap::Get in unit tests
-#ifndef CONFIG_BUILD_FOR_HOST_UNIT_TEST
+#if !CONFIG_BUILD_FOR_HOST_UNIT_TEST
     success = (Attributes::FeatureMap::Get(kRootEndpointId, &featureMap) == EMBER_ZCL_STATUS_SUCCESS);
 #endif
     return success ? ((featureMap & to_underlying(Feature::kCheckInProtocolSupport)) != 0) : false;
 }
 
+void ICDManager::SendCheckInMsgs()
+{
+    VerifyOrDie(mStorage != nullptr);
+    VerifyOrDie(mFabricTable != nullptr);
+    for (const auto & fabricInfo : *mFabricTable)
+    {
+        uint16_t supported_clients = IcdManagementServer::GetInstance().GetClientsSupportedPerFabric();
+        // TODO retrieve Check-in counter
+        // SecureChannel::CounterType counter = GetCheckInCounter();
+
+        ChipLogProgress(AppServer, "Max clients ICD per fabric %d", supported_clients);
+
+        IcdMonitoringTable table(*mStorage, fabricInfo.GetFabricIndex(), supported_clients /*Table entry limit*/);
+        if (!table.IsEmpty())
+        {
+            ChipLogProgress(AppServer, "Retrieving ICD Check-in Clients");
+            for (uint16_t i = 0; i < table.Limit(); i++)
+            {
+                IcdMonitoringEntry entry;
+                if (CHIP_NO_ERROR != table.Get(i, entry))
+                {
+                    // No more entries in the table
+                    break;
+                }
+
+                // TODO check for active subscription
+
+                // bool active = InteractionModelEngine::GetInstance()->IsSubscriptionActive(entry.fabricIndex, entry.checkInNodeID,
+                // entry.monitoredSubject); if (!active)
+                // {
+                //     // TODO After PR 29260 is merged.
+                //     ByteSpan appData;
+                //     uint8_t b[CheckinMessage::sMinPayloadSize] = { 0 };
+                //     MutableByteSpan output{ b };
+
+                //     // Prepare Check-in payload
+                //     if(CHIP_NO_ERROR != CheckinMessage::GenerateCheckinMessagePayload(  entry.key,
+                //                                                     counter,
+                //                                                     appData,
+                //                                                     output))
+                //     {
+                //         ChipLogError(AppServer, "skipping node %lu", entry.checkInNodeID);
+                //         continue;
+                //     }
+
+                //     // Todo send this somewhere so that the address can be resolved and
+                //     // the payload can be send throught the network interface
```

Comment:
```suggestion
                //     // the payload can be sent through the network interface
```

You're going to need to do dns-sd discovery, which is async, etc...  Should that be done _before_ creating the payloads, so we don't have to carry the payloads around in memory?

Reply:
I've not tackle the dns-sd discovery yet. but most probably.